internal/experiment/openvpn/targets.go

```diff
@@ -1,49 +1,170 @@
+	resolver := netx.NewResolver(netx.Config{
+		BogonIsError: false,
+		Logger:       logger,
+		Saver:        nil,
+	})
+
+	addrs := []string{}
+
+	var lastErr error
+
+	// get the set of all IPs for all the hostnames we have.
+	for _, hostname := range defaultOONIEndpoints {
+		resolved, err := lookupHost(context.Background(), hostname, resolver)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		for _, ipaddr := range resolved {
+			if !slices.Contains(addrs, ipaddr) {
+				addrs = append(addrs, ipaddr)
+			}
+		}
+	}
+
+	// Sample a max of maxDefaultOONIAddresses
+
+	sampled := sampleN(addrs, maxDefaultOONIAddresses)
+
+	// Remove the bogons
+
+	valid := []string{}
+
+	for _, addr := range sampled {
+		if !netxlite.IsBogon(addr) {
+			valid = append(valid, addr)
+		}
+	}
+
+	// We only return error if the filtered list is zero len.
+
+	if (len(valid) == 0) && (lastErr != nil) {
+		return valid, lastErr
 	}
-	return defaultOONITargetURL(ip) + "&transport=udp", nil
+
+	return valid, nil
 }
 
-func defaultOONIOpenVPNTargetTCP() (string, error) {
-	ip, err := resolveTarget(defaultOpenVPNEndpoint)
-	if err != nil {
-		return "", err
+func lookupHost(ctx context.Context, hostname string, r model.Resolver) ([]string, error) {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+	return r.LookupHost(ctx, hostname)
+}
+
+// pickOONIOpenVPNTargets crafts targets from the passed array of IP addresses.
+func pickOONIOpenVPNTargets(ipaddrList []string) ([]string, error) {
+
+	// Step 1. Create endpoint list.
+
+	endpoints := []endpoint{}
+	for _, ipaddr := range ipaddrList {
+
+		// Probe the canonical 1194/udp and 1194/tcp ports
+
+		endpoints = append(endpoints, endpoint{
+			Obfuscation: "none",
+			Port:        "1194",
+			Protocol:    "openvpn",
+			Provider:    "oonivpn",
+			IPAddr:      ipaddr,
+			Transport:   "tcp",
+		})
+		endpoints = append(endpoints, endpoint{
+			Obfuscation: "none",
+			Port:        "1194",
+			Protocol:    "openvpn",
+			Provider:    "oonivpn",
+			IPAddr:      ipaddr,
+			Transport:   "udp",
+		})
+
+	}
+
+	// Pick one IP from the list and sample on non-standard ports
+	// to check if the standard port was filtered.
+
+	extra := ipaddrList[rand.Intn(len(ipaddrList))]
+	endpoints = append(endpoints, endpoint{
+		Obfuscation: "none",
+		Protocol:    "openvpn",
+		Provider:    "oonivpn",
+		IPAddr:      extra,
+		Port:        "53",
+		Transport:   "udp",
+	})
+	endpoints = append(endpoints, endpoint{
+		Obfuscation: "none",
+		Protocol:    "openvpn",
+		Provider:    "oonivpn",
+		IPAddr:      extra,
+		Port:        "443",
+		Transport:   "tcp",
+	})
+
+	// Step 2. Create targets for the selected endpoints.
+
+	targets := make([]string, 0)
+	for _, e := range endpoints {
+		targets = append(targets, e.AsInputURI())
+	}
+	if len(targets) > 0 {
+		return targets, nil
 	}
-	return defaultOONITargetURL(ip) + "&transport=tcp", nil
+	return nil, fmt.Errorf("cannot find any usable endpoint")
```

Comment:
Will this result in a hard failure of the test? Is there some way we can ensure that's not the case? For example we might want for certain runs in particular countries to not run the vpn tests and it would be ideal if that results in a noop as opposed to the test being considered a failure.

Reply:
if this function returns error, we simply return an empty targets list, see: 
https://github.com/ooni/probe-cli/blob/d9c4b3e4e8ea764829e7e0f5cf5f2a302bdf8b61/internal/experiment/openvpn/richerinput.go#L113

Reply:
see the [comment below](https://github.com/ooni/probe-cli/pull/1654#issuecomment-2436080116)